### PR TITLE
ISSUE #2785 - Federations page

### LINKED ISSUE #2785: Revamping - Federations page
### Description
Federations will be reimplemented with the new style. It will provide the functionality of viewing all the federations from a project. Be able to upload/edit/delete them. Provide some sort of search. 
It will also
The list will be based of the containers lists components as well as components specially needed for the federations.

### Goals
- [x] To have an all federations list based on the selected project
- [x] To have a favourites federations list based on the selected project
- [x] To add and remove federations from the the favourite list
- [x] To have blank states when the lists are empty
- [ ] Show revisions detailes when a row has been clicked

### Tasks
- [ ] Use lingui \<Trans id=".." /> tag for translations for the columns and blank state
- [ ] Use lingui i18n._('...') for translations when you can only pass strings 
- [ ] Create a reusable list item for federations using the components like \<dashboardlistitem ...> etc. 
- [ ] Use that list for all federations and favourites
- [ ] Use the columns for sorting the lists (if this bit goes into redux/selectors it depends on wether we want to keep the state when navigating throught the app) 
- [ ] Add a blankstate already created for containers when there are no items in the list (remember that no button for creating federations is needed at this point)
- [ ] Add api endpoint for fetching the federations to the api service ( I mean in the frontend)
- [ ] Add api endpoint for Adding/removing a federation as favourite the api service
- [ ] Add saga fetch actions and selectors for fetching and selecting the federations
- [ ] Add saga add/remove action for adding/removing the federation as favourite
- [ ] Use createHookSelectors for the the selectors as exemplified in **src\v5\services\selectorsHooks\currentUserSelectors.hooks.ts**
- [ ] Use actiondispatcher as exemplified in  **src\v5\services\actionsDispatchers\currentUsersActions.dispatchers.ts**

### Related Resources
MockUp : https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-(live)?node-id=2427%3A36135

### PR DESCRIPTION
**Pull request by EmilTheSadCat**
This fixes #2785

test URL: https://issue-2785.dev.3drepo.io/v5/dashboard/carmen/843ba9f0-00c6-11ec-8d97-41a278fb55fd/t/federations

### CODE REVIEW

**Comment by carmenfan** [comment]
@EmilTheSadCat there's now a conflict. can you take a look? 

Though it might be better to fix the conflicts on https://github.com/3drepo/3drepo.io/pull/2831 first so we can merge that in , just in case there's more!

**Comment by george-hadfield** [comment]
### Close search icon not in correct place
Icon needs more padding to the right
 
### Pictures & Links
image

**Comment by george-hadfield** [comment]
### Move 'New Federation' button (large one) to inside the 'All Federations' list
It would be better if the behaviour of this large button was that it doesn't show if the 'All Federations' list is closed. It's purpose is for those users that are at the bottom of a very long Federations list. If it is closed, then there are buttons available at the top of the screen. Could this also be applied to the Containers section too?
 
### Pictures & Links
 
image

**Review by george-hadfield — CHANGES_REQUESTED** [review]
Left two small issues. Other than that, everything seems good!

**Comment by george-hadfield** [comment]
### Empty search on favourites and All Federations displays incorrect text information
Currently it filters out to say you don't have any favourites or any Federations if there are no results to display.

It should display "We couldn't find a match for **“Search term”.** Please try another search." 

image

**Comment by sanmont3drepo** [comment]
I've changed the base of the PR to staging due to ISSUE #2796  already been merged in. So there's some conlicts that must be solved.

**Comment by EmilTheSadCat** [comment]
@sanmont3drepo Conflicts resolved.

@george-hadfield I've fixed the issues you've mentioned above. It's ready for another review.

**Comment by george-hadfield** [comment]
@EmilTheSadCat just check and everything is working fine now. Thank you!

File: frontend/src/v5/ui/routes/dashboard/projects/federations/federationsList/federationsList.component.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+					<DashboardListHeaderLabel name="lastUpdated" width={180}>
+						<Trans id="federations.list.header.lastUpdated" message="Last updated" />
+					</DashboardListHeaderLabel>
+				</DashboardListHeader>
+				{useMemo(() => (
+					<DashboardList>
+						{!isEmpty(sortedList) ? (
+							sortedList.map((federation, index) => (
+								federation.hasStatsPending ? (
+									<SkeletonListItem delay={index / 10} key={federation._id} />
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/federations/federationsList/federationsList.component.tsx:139]: I think we could have this inside of fedeationListItem, as it is it responsability to represent a ui item of a particular federation, including if it is fully loaded. Same goes with containers. If you think this doesnt seem like a good idea, let me know.

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Comment by carmenfan** [comment]
@m1chalkubiak can you take a look at the changes here before you move onto the upload file dialog?

It also has a bit of conflicts with staging now

**Comment by m1chalkubiak** [comment]
> @m1chalkubiak can you take a look at the changes here before you move onto the upload file dialog?
> 
> It also has a bit of conflicts with staging now

Sure @carmenfan I will handle it.

File: frontend/src/v5/ui/components/dashboard/dashboardList/dashboardSkeletonList/dashboardSkeletonList.component.tsx [hunk 2]
```diff
@@ -31,14 +29,15 @@ export const DashboardSkeletonList = ({ itemsAmount = 10, itemComponent }: IDash
 	return (
 		<DashboardList>
 			{list.map((item, index) => (
-				<div key={uuid()}>
+				<>
 					{
 						React.cloneElement(itemComponent, {
+							key: itemComponent.props.name,
```
[the remark this answers is not in the document]
Comment (m1chalkubiak) [reply frontend/src/v5/ui/components/dashboard/dashboardList/dashboardSkeletonList/dashboardSkeletonList.component.tsx:35]: I've fixed this. In fact, there were two issues here, one was the value of key which You absolutely had right, but also the key should be placed for direct children of rendered element list which, in this case, is the Fragment element.

**Comment by m1chalkubiak** [comment]
I am still working on resolving conflicts because there are quite a lot of them, but I can already see that when this is finished I will have to improve the translations in all the newly created components in this PR because they used `lingui`. I will have to review them and correct it by one by one.

**Comment by sanmont3drepo** [comment]
> I am still working on resolving conflicts because there are quite a lot of them, but I can already see that when this is finished I will have to improve the translations in all the newly created components in this PR because they used `lingui`. I will have to review them and correct it by one by one.

Is pretty straight forward to do so. just use `<FormattedMessage  />` instead of  `<Trans />` and `defaultMessage='..'` instead of `message='..'`` 

For rich text you can see an example here:
https://github.com/3drepo/3drepo.io/blob/e61ac076d49feb5d9793d8058f2c0596e54f437e/frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/emptySearchResults/emptySearchResults.component.tsx#L28-L36

**Comment by sanmont3drepo** [comment]
> > I am still working on resolving conflicts because there are quite a lot of them, but I can already see that when this is finished I will have to improve the translations in all the newly created components in this PR because they used `lingui`. I will have to review them and correct it by one by one.
> 
> Is pretty straight forward to do so. just use `<FormattedMessage />` instead of `<Trans />` and `defaultMessage='..'` instead of `message='..'``
> 
> For rich text you can see an example here:
> 
> https://github.com/3drepo/3drepo.io/blob/e61ac076d49feb5d9793d8058f2c0596e54f437e/frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/emptySearchResults/emptySearchResults.component.tsx#L28-L36

For sagas or parts of the code  that doesnt have react we should use `formatMessage` from 'services/intl'
look at:

https://github.com/3drepo/3drepo.io/blob/e61ac076d49feb5d9793d8058f2c0596e54f437e/frontend/src/v5/store/projects/projects.sagas.ts#L21-L30

**Comment by m1chalkubiak** [comment]
> > I am still working on resolving conflicts because there are quite a lot of them, but I can already see that when this is finished I will have to improve the translations in all the newly created components in this PR because they used `lingui`. I will have to review them and correct it by one by one.
> 
> Is pretty straight forward to do so. just use `<FormattedMessage />` instead of `<Trans />` and `defaultMessage='..'` instead of `message='..'``
> 
> For rich text you can see an example here:
> 
> https://github.com/3drepo/3drepo.io/blob/e61ac076d49feb5d9793d8058f2c0596e54f437e/frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/emptySearchResults/emptySearchResults.component.tsx#L28-L36

@sanmont3drepo I did not say that it will be complicated, but only time-consuming, and as You know, I don't have much left. And I would like to devote as much as possible to things that will bring some value to the project.

Anyway, it is done and it seems to me that everything is working fine. I also figured out what was the issue with the `TextOverflow` component I corrected it here and I will add it also at https://github.com/3drepo/3drepo.io/pull/2852

**Review by sanmont3drepo — APPROVED** [review]

**Comment by carmenfan** [comment]
@george-hadfield  can you take another look on this one?

**Comment by Charence** [comment]
@3drepo/product-team Please check this

**Comment by carmenfan** [comment]
@george-hadfield  the link should work again now

**Comment by carmenfan** [comment]
@sanmont3drepo From the last design meeting, It's been concluded that they don't need categories in Federations so there has been a change on what is shown on the page 
image

The code is now where the category used to be, and the description is added below the model name

**Comment by george-hadfield** [comment]
### Remove hover state from Federation and Container items
Currently a shadow and scale when a user hovers the item row, however, after seeing this implemented, it's too much movement. I'd like to remove the shadow and scale on hover if possible.
 
### Pictures & Links

image

**Comment by george-hadfield** [comment]
### Responsive: Content needs margin left and right at smaller breakpoints. 
Please refer to the containers responsive: https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-(live)?node-id=3274%3A46976

### Pictures & Links

Build
image

Design
image

**Comment by george-hadfield** [comment]
### Responsive: Columns should scale the same as containers
Please refer to the Containers responsive design for guidance https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-(live)?node-id=3274%3A46976

### Pictures & Links 

Build
image

**Comment by george-hadfield** [comment]
### Main nav needs to use the updated V2 style

### Pictures and links

Build
image

Design
image

**Comment by sanmont3drepo** [comment]
> ### Main nav needs to use the updated V2 style
> ### Pictures and links
> Build image
> 
> Design image

This will be addressed in https://github.com/3drepo/3drepo.io/pull/2852

**Comment by sanmont3drepo** [comment]
> ### Responsive: Content needs margin left and right at smaller breakpoints.
> Please refer to the containers responsive: https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-(live)?node-id=3274%3A46976
> 
> ### Pictures & Links
> Build image
> 
> Design image

I believe this needs to be addressed in https://github.com/3drepo/3drepo.io/pull/2852  (it has all the new code regarding responsiveness)

**Comment by sanmont3drepo** [comment]
> ### Remove hover state from Federation and Container items
> Currently a shadow and scale when a user hovers the item row, however, after seeing this implemented, it's too much movement. I'd like to remove the shadow and scale on hover if possible.
> 
> ### Pictures & Links
> image

Will be addressed in https://github.com/3drepo/3drepo.io/issues/2903

**Review by george-hadfield — APPROVED** [review]
Approved based on Carmens comments in Teams

**Review by sanmont3drepo — APPROVED** [review]

**Pull request merged**